core/validator_manager.go

```diff
@@ -0,0 +1,137 @@
+package core
+
+import (
+	"bytes"
+	"errors"
+	"math/big"
+	"sync"
+
+	"github.com/0xPolygon/go-ibft/messages/proto"
+)
+
+var (
+	errVotingPowerNotCorrect = errors.New("total voting power is zero or less")
+)
+
+// ValidatorManager keeps voting power and other informations about validators
+type ValidatorManager struct {
+	vpLock *sync.RWMutex
+
+	// quorumSize represents quorum for the height specified in the current View
+	quorumSize *big.Int
+
+	// validatorsVotingPower is a map of the validator addresses on their voting power for
+	// the height specified in the current View
+	validatorsVotingPower map[string]*big.Int
+
+	log Logger
+}
+
+// NewValidatorManager creates new ValidatorManager
+func NewValidatorManager(log Logger) *ValidatorManager {
+	return &ValidatorManager{
+		quorumSize:            big.NewInt(0),
+		validatorsVotingPower: nil,
+		log:                   log,
+		vpLock:                &sync.RWMutex{},
+	}
+}
+
+// Init sets voting power and quorum size
+func (vm *ValidatorManager) Init(validatorsVotingPower map[string]*big.Int) error {
+	vm.vpLock.Lock()
+	defer vm.vpLock.Unlock()
+
+	totalVotingPower := calculateTotalVotingPower(validatorsVotingPower)
+	if totalVotingPower.Cmp(big.NewInt(0)) <= 0 {
+		return errVotingPowerNotCorrect
+	}
+
+	vm.validatorsVotingPower = validatorsVotingPower
+	vm.quorumSize = calculateQuorum(totalVotingPower)
+
+	return nil
+}
+
+// HasQuorum provides information on whether messages have reached the quorum
+func (vm *ValidatorManager) HasQuorum(addressMap map[string]struct{}) bool {
```

Comment:
Why do we need a map here?

Reply:
to insure that `ValidatorManager` does not receive the same validator twice.